### PR TITLE
Assignment2

### PR DESCRIPTION
**Pull request by NPThompson**

### CODE REVIEW

File: Logger/FileLogger.cs [hunk 1]
```diff
@@ -0,0 +1,16 @@
+using System.IO;
+
+namespace Logger
+{
+    public class FileLogger : BaseLogger
+    {
+        public string filePath {get; set;}
```
Comment (mmwoodfo) [inline Logger/FileLogger.cs:7]: There is no reason for this to be public - it should be kept a private member of the class.
```suggestion
        private string _filePath;
```

Also, this line would be giving you a warning if nullability was enabled as it is a possible null reference.

File: Logger/FileLogger.cs [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+        public override void Log(LogLevel logLevel, string message)
+        {
+            StreamWriter LogFile = File.AppendText(filePath);
```
Comment (mmwoodfo) [inline Logger/FileLogger.cs:11]: Another way to accomplish this is with the [using ](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/using-statement) statement.
```suggestion
            using(streamWriter streamWriter = new StreamWriter(filePath, true))
            {
                 streamWriter.WriteLine("Log message stuff here...");
            }
```

File: Logger/BaseLoggerMixins.cs [hunk 3]
```diff
@@ -2,6 +2,33 @@
 {
     public static class BaseLoggerMixins
     {
-        
+        public static void Error(this BaseLogger logger, string message)
```
Comment (mmwoodfo) [inline Logger/BaseLoggerMixins.cs:5]: Is there a reason there are two `Error` extension methods?

File: Logger/BaseLoggerMixins.cs [hunk 4]
```diff
@@ -2,6 +2,33 @@
...
+            if(logger == null)
+            throw new System.ArgumentNullException();
+
+            logger.Log(LogLevel.Error, message);
+        }
+
+        //Allows one to write: MyLogger.Error("Some message {0} {1}", 12345, 67890 );
+        public static void Error(this BaseLogger logger, string formatMessage, params object[] insertIntoMessage)
+        {
+            logger.Log(LogLevel.Error, string.Format(formatMessage, insertIntoMessage));
```
Comment (mmwoodfo) [inline Logger/BaseLoggerMixins.cs:16]: This should be checking if the `BaseLogger` is null before using it to log a message. The same goes for the below extension methods.

File: Logger/BaseLogger.cs [hunk 5]
```diff
@@ -2,6 +2,8 @@
 {
     public abstract class BaseLogger
     {
+        public string className {get; set;}
```
Comment (mmwoodfo) [inline Logger/BaseLogger.cs:5]:
```suggestion
        public string ClassName {get; set;}
```

File: Logger.Tests/LogFactoryTests.cs [hunk 6]
```diff
@@ -6,6 +6,25 @@ namespace Logger.Tests
...
     {
-        
+        [TestMethod]
+        public void LogFactory_SetsClassName()
+        {
+         // Arrange
+         LogFactory myLogFactory = new LogFactory();
+         myLogFactory.configureFileLogger("foobar");
+
+         FileLogger myLogger = (FileLogger)myLogFactory.CreateLogger(nameof(LogFactoryTests));
```
Comment (mmwoodfo) [inline Logger.Tests/LogFactoryTests.cs:16]: Looks great! No need to cast to a file logger, you can simply create a new BaseLogger.
```suggestion
         BaseLogger myLogger = myLogFactory.CreateLogger(nameof(LogFactoryTests));
```

File: Logger.Tests/LogFactoryTests.cs [hunk 7]
```diff
@@ -6,6 +6,25 @@ namespace Logger.Tests
...
+                  
+         // Test
+         Assert.AreEqual(myLogger.className, nameof(LogFactoryTests));
+        }
+
+        [TestMethod]
+        public void LogFactory_ReturnsNull_IfFileLoggerNotConfigured()
+        {
+            FileLogger myLogger = (FileLogger)(new LogFactory()).CreateLogger("FileLogger");
```
Comment (mmwoodfo) [inline Logger.Tests/LogFactoryTests.cs:25]:
```suggestion
            LogFactory logFactory = new LogFactory();
```

We are testing the `LogFactory` class, so you should be creating a new `FileLogger` via the `LogFactory` class method `CreateLogger`.

File: Logger.Tests/LogFactoryTests.cs [hunk 8]
```diff
@@ -6,6 +6,25 @@ namespace Logger.Tests
...
+         myLogFactory.configureFileLogger("foobar");
+
+         FileLogger myLogger = (FileLogger)myLogFactory.CreateLogger(nameof(LogFactoryTests));
+                  
+         // Test
+         Assert.AreEqual(myLogger.className, nameof(LogFactoryTests));
+        }
+
+        [TestMethod]
+        public void LogFactory_ReturnsNull_IfFileLoggerNotConfigured()
```
Comment (mmwoodfo) [inline Logger.Tests/LogFactoryTests.cs:23]: Another test you could have included would be to check the opposite of this test - that the logger is not null when configured properly.

File: Logger.Tests/LogFactoryTests.cs (hunk 7 above)
Comment (mmwoodfo) [reply Logger.Tests/LogFactoryTests.cs:25]: You would then need an extra line of code that creates a new BaseLogger - similar to what you did above on line 16.

File: Logger.Tests/FileLoggerTests.cs [hunk 9]
```diff
@@ -1,10 +1,35 @@
...
     public class FileLoggerTests
     {
+        [TestMethod]
+        public void FileLogger_AppendsLogsToFile()
+        {
+            string message1 = "What is the average air speed velocity of an unladen swallow?";
+            string message2 = "African or European?";
+
+            if(File.Exists("fileLoggerTest.txt"))
+                File.Delete("fileLoggerTest.txt");
```
Comment (mmwoodfo) [inline Logger.Tests/FileLoggerTests.cs:16]:
```suggestion
                File.Delete(fileName);
```

An even better suggestion would be to use the `GetRandomFileName()` method to generate a file name and rather than cleaning up after a previous unit test - you can delete the file at the end of the test that way you don't have extra data lying around after a test has finished running. (Just be sure to wrap the `File.Delete(...)` method in a finally block so it gets removed even if the unit test fails.)

File: Logger.Tests/FileLoggerTests.cs [hunk 10]
```diff
@@ -1,10 +1,35 @@
...
     [TestClass]
     public class FileLoggerTests
     {
+        [TestMethod]
+        public void FileLogger_AppendsLogsToFile()
+        {
+            string message1 = "What is the average air speed velocity of an unladen swallow?";
+            string message2 = "African or European?";
+
+            if(File.Exists("fileLoggerTest.txt"))
```
Comment (mmwoodfo) [inline Logger.Tests/FileLoggerTests.cs:15]:
```suggestion
            string fileName = "fileLoggerTest.txt"
            if(File.Exists(fileName))
```
Since you are typing this multiple times, it may be better to store it in a variable you then reuse.

File: Logger.Tests/FileLoggerTests.cs [hunk 11]
```diff
@@ -1,10 +1,35 @@
...
+                File.Delete("fileLoggerTest.txt");
         
+            LogFactory logFactory = new LogFactory();
```
Comment (mmwoodfo) [inline Logger.Tests/FileLoggerTests.cs:18]: Since we are now testing the `FileLogger` class - it is best to avoid using the `LogFactory` class to create your file logger since the `LogFactory` is not the current system under test. Instead, you can directly instantiate your `FileLogger`

File: Logger.Tests/FileLoggerTests.cs [hunk 12]
```diff
@@ -1,10 +1,35 @@
...
+            logFactory.configureFileLogger("fileLoggerTest.txt");
+
+            FileLogger fileLogger = (FileLogger)logFactory.CreateLogger(nameof(FileLoggerTests));
+
+            fileLogger.Log(LogLevel.Warning, message1);
+            fileLogger.Log(LogLevel.Warning, message2);
+
+            StreamReader testFileLoggerReader = new StreamReader("fileLoggerTest.txt");
```
Comment (mmwoodfo) [inline Logger.Tests/FileLoggerTests.cs:26]: An alternative option:
```C#
List<string> lines = File.ReadLines(filePath).ToList();
```

File: Logger.Tests/BaseLoggerMixinsTests.cs [hunk 13]
```diff
@@ -26,14 +26,59 @@ public void Error_WithData_LogsMessage()
...
-            //logger.Error("Message {0}", 42);
+            logger.Error("Message {0}", 42);
 
             // Assert
             Assert.AreEqual(1, logger.LoggedMessages.Count);
             Assert.AreEqual(LogLevel.Error, logger.LoggedMessages[0].LogLevel);
             Assert.AreEqual("Message 42", logger.LoggedMessages[0].Message);
         }
 
+        [TestMethod]
```
Comment (mmwoodfo) [inline Logger.Tests/BaseLoggerMixinsTests.cs:37]: You're missing the ExpectedException unit tests for the Warning, Information, and Debug extension methods.

**Review by mmwoodfo — COMMENTED** [review]
##  (NPThompson)

## Assignment2

| Build/Compile (Build succeeded: 0 Error(s), 0 Warning(s))| **<10>** |
| --------------------- | ------ |
| Assignment compiles even though major pieces might be missing, or there are significant problems: **✔** |  **5** |


| Assignment  | **<10 (missing lots) - 20 (minor)>** |
| --------------------- | ------ |
| <Project Assignment Goes Here> | ❌✔ |


| Project Analysis |   |
| -----------------| - |
| Disabled Warnings **** | **✔**
| Nullable Enabled: | **❌** |
| Appropriately handle all warnings | **✔** |
| CodeReview Performed | **✔** |
| Add the code analysis packages and appropriately handle all warnings* |  **❌** |
*<sub><sup>`IntelliTectAnalyzer.dll`,`Microsoft.NetCore.Analyzers.dll`,`Microsoft.CodeQuality.Analyzers.dll`,`Microsoft.NetCore.Analyzers.dll`,`Microsoft.NetFramework.Analyzers.dll`. (It is okay to appropriately disable rules such as  localization/culture rules.)</sub></sup>


| ProjectName | Nullable | Solutions | Analyzers | Unexpected Disabled Warnings | Waning Level |
| -- | -- | -- | -- | -- | -- |
| Logger | Disable | Logger.sln |  |  | 4 |
| Logger.Tests | Enable | Logger.sln |  |  | 4 |
| LoginStuff | Disable |  |  |  | 4 |
| LoginStuff.Tests | Disable |  |  |  | 4 |


| Testing | **<0 - 20>** |
| --------------------- | ------ |
| Test Run Succeeded: Total tests=8, Passed=8, Failed=0, Skipped=0 | **96%** |
| Unit test project fails to compile, most test fails, or contains significant problems. | **<10>** |
| Most assigned work is tested or contain some errors. | **<5>** |
| All assigned work is tested | **<5>** |


| Extra Credit | **<0 - 10>** |
| --------------------- | ------ |
| <Extra Credit Goes Here> | ❌✔ |

File: Logger/LogFactory.cs [hunk 14]
```diff
@@ -2,10 +2,18 @@
 {
     public class LogFactory
     {
-        public BaseLogger CreateLogger(string className)
+        private string _fileLoggerFilePath;
```
Comment (mmwoodfo) [inline Logger/LogFactory.cs:5]: _FileLoggerFilePath
